From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zml <zml@stellardrift.ca>
Date: Wed, 31 Mar 2021 16:12:46 -0700
Subject: [PATCH] Make methods hold a reference to their declaring class

This is used in several later commits.

diff --git a/src/org/jetbrains/java/decompiler/struct/StructMethod.java b/src/org/jetbrains/java/decompiler/struct/StructMethod.java
index f58ee71d23ded6d1fe495b5872bd7965b4c6e61e..4e5c7b4da350020f4f69da956eda4384b86df2aa 100644
--- a/src/org/jetbrains/java/decompiler/struct/StructMethod.java
+++ b/src/org/jetbrains/java/decompiler/struct/StructMethod.java
@@ -41,7 +41,7 @@ public class StructMethod extends StructMember {
       attributes.putAll(code.codeAttributes);
     }
 
-    return new StructMethod(accessFlags, attributes, values[0], values[1], bytecodeVersion, own ? code : null);
+    return new StructMethod(accessFlags, attributes, values[0], values[1], bytecodeVersion, own ? code : null, clQualifiedName);
   }
 
   private static final int[] opr_iconst = {-1, 0, 1, 2, 3, 4, 5};
@@ -57,13 +57,15 @@ public class StructMethod extends StructMember {
   private final int codeFullLength;
   private InstructionSequence seq = null;
   private boolean expanded = false;
+  private final String classQualifiedName;
 
   private StructMethod(int accessFlags,
                        Map<String, StructGeneralAttribute> attributes,
                        String name,
                        String descriptor,
                        int bytecodeVersion,
-                       StructCodeAttribute code) {
+                       StructCodeAttribute code,
+                       String classQualifiedName) {
     super(accessFlags, attributes);
     this.name = name;
     this.descriptor = descriptor;
@@ -76,6 +78,7 @@ public class StructMethod extends StructMember {
     else {
       this.localVariables = this.codeLength = this.codeFullLength = -1;
     }
+    this.classQualifiedName = classQualifiedName;
   }
 
   public void expandData(StructClass classStruct) throws IOException {
@@ -378,4 +381,8 @@ public class StructMethod extends StructMember {
   public String toString() {
     return name;
   }
+
+  public String getClassQualifiedName() {
+    return classQualifiedName;
+  }
 }
