From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zml <zml@stellardrift.ca>
Date: Wed, 31 Mar 2021 16:12:46 -0700
Subject: [PATCH] Make methods hold a reference to their declaring class

This is used in several later commits.

diff --git a/src/org/jetbrains/java/decompiler/struct/StructMethod.java b/src/org/jetbrains/java/decompiler/struct/StructMethod.java
index 2336000ea054e5f2e0ef88d30f4a7bc8ee04494f..3b14d7fc755a2d9c33a022aaa2ab728c4a12a4e8 100644
--- a/src/org/jetbrains/java/decompiler/struct/StructMethod.java
+++ b/src/org/jetbrains/java/decompiler/struct/StructMethod.java
@@ -39,7 +39,7 @@ public class StructMethod extends StructMember {
       attributes.putAll(code.codeAttributes);
     }
 
-    return new StructMethod(accessFlags, attributes, values[0], values[1], bytecodeVersion, own ? code : null);
+    return new StructMethod(accessFlags, attributes, values[0], values[1], bytecodeVersion, own ? code : null, clQualifiedName);
   }
 
   private static final int[] opr_iconst = {-1, 0, 1, 2, 3, 4, 5};
@@ -55,13 +55,15 @@ public class StructMethod extends StructMember {
   private final int codeFullLength;
   private InstructionSequence seq = null;
   private boolean expanded = false;
+  private final String classQualifiedName;
 
   private StructMethod(int accessFlags,
                        Map<String, StructGeneralAttribute> attributes,
                        String name,
                        String descriptor,
                        int bytecodeVersion,
-                       StructCodeAttribute code) {
+                       StructCodeAttribute code,
+                       String classQualifiedName) {
     super(accessFlags, attributes);
     this.name = name;
     this.descriptor = descriptor;
@@ -74,6 +76,7 @@ public class StructMethod extends StructMember {
     else {
       this.localVariables = this.codeLength = this.codeFullLength = -1;
     }
+    this.classQualifiedName = classQualifiedName;
   }
 
   public void expandData(StructClass classStruct) throws IOException {
@@ -371,4 +374,8 @@ public class StructMethod extends StructMember {
   public String toString() {
     return name;
   }
+
+  public String getClassQualifiedName() {
+    return classQualifiedName;
+  }
 }
