From 4c12c7a453ab63d9b966d8247ed88e4873fe7d29 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Sun, 2 Aug 2015 19:22:15 -0700
Subject: Fix tests for inner class sorting.


diff --git a/testData/results/TestAnonymousClass.dec b/testData/results/TestAnonymousClass.dec
index 75d1df239bb56bc0a4be42a1310957721146ece2..fdce11007ba778eaaeb4081e1a11041b2573feca 100644
--- a/testData/results/TestAnonymousClass.dec
+++ b/testData/results/TestAnonymousClass.dec
@@ -67,15 +67,8 @@ public abstract class TestAnonymousClass {
       boolean var1 = true;// 39
    }// 40
 
-   static class InnerRecursive {
-      TestAnonymousClass.InnerRecursive r;
-
-      public InnerRecursive(TestAnonymousClass.InnerRecursive var1) {
-         this.r = var1;// 105
-      }// 106
-
-      void foo() {
-      }// 110
+   interface I {
+      void foo() throws Exception;
    }
 
    private static class Inner {
@@ -87,8 +80,15 @@ public abstract class TestAnonymousClass {
       };
    }
 
-   interface I {
-      void foo() throws Exception;
+   static class InnerRecursive {
+      TestAnonymousClass.InnerRecursive r;
+
+      public InnerRecursive(TestAnonymousClass.InnerRecursive var1) {
+         this.r = var1;// 105
+      }// 106
+
+      void foo() {
+      }// 110
    }
 }
 
@@ -180,24 +180,24 @@ class 'pkg/TestAnonymousClass' {
    }
 }
 
+class 'pkg/TestAnonymousClass$Inner$1' {
+   method 'run ()V' {
+      0      76
+      1      76
+      2      77
+      3      77
+      4      78
+   }
+}
+
 class 'pkg/TestAnonymousClass$InnerRecursive' {
    method '<init> (Lpkg/TestAnonymousClass$InnerRecursive;)V' {
-      6      73
-      9      74
+      6      86
+      9      87
    }
 
    method 'foo ()V' {
-      0      77
-   }
-}
-
-class 'pkg/TestAnonymousClass$Inner$1' {
-   method 'run ()V' {
-      0      83
-      1      83
-      2      84
-      3      84
-      4      85
+      0      90
    }
 }
 
@@ -223,9 +223,9 @@ Lines mapping:
 53 <-> 18
 54 <-> 19
 55 <-> 20
-66 <-> 84
-67 <-> 85
-68 <-> 86
+66 <-> 77
+67 <-> 78
+68 <-> 79
 75 <-> 24
 76 <-> 25
 77 <-> 26
@@ -234,9 +234,9 @@ Lines mapping:
 91 <-> 37
 92 <-> 38
 93 <-> 39
-105 <-> 74
-106 <-> 75
-110 <-> 78
+105 <-> 87
+106 <-> 88
+110 <-> 91
 Not mapped:
 18
 104
diff --git a/testData/results/TestInnerLocal.dec b/testData/results/TestInnerLocal.dec
index 6ca3a7e1053087d91543ce588eb81e19c9f3432d..e5aba7da20b0b57a4c796a36c24960f38c6eac31 100644
--- a/testData/results/TestInnerLocal.dec
+++ b/testData/results/TestInnerLocal.dec
@@ -28,6 +28,14 @@ public class TestInnerLocal {
       new TestInnerLocal.Inner1Static.Inner2Static("test");// 40
    }// 41
 
+   class Inner1 {
+      final String x;
+
+      public Inner1(String var2) {
+         this.x = var2;// 46
+      }// 47
+   }
+
    static class Inner1Static {
       final String x;
 
@@ -43,14 +51,6 @@ public class TestInnerLocal {
          }// 60
       }
    }
-
-   class Inner1 {
-      final String x;
-
-      public Inner1(String var2) {
-         this.x = var2;// 46
-      }// 47
-   }
 }
 
 class 'TestInnerLocal$1Inner' {
@@ -84,24 +84,24 @@ class 'TestInnerLocal$2Inner' {
    }
 }
 
-class 'TestInnerLocal$Inner1Static' {
-   method '<init> (Ljava/lang/String;)V' {
-      6      34
-      9      35
+class 'TestInnerLocal$Inner1' {
+   method '<init> (LTestInnerLocal;Ljava/lang/String;)V' {
+      b      34
+      e      35
    }
 }
 
-class 'TestInnerLocal$Inner1Static$Inner2Static' {
+class 'TestInnerLocal$Inner1Static' {
    method '<init> (Ljava/lang/String;)V' {
-      6      41
-      9      42
+      6      42
+      9      43
    }
 }
 
-class 'TestInnerLocal$Inner1' {
-   method '<init> (LTestInnerLocal;Ljava/lang/String;)V' {
-      b      50
-      e      51
+class 'TestInnerLocal$Inner1Static$Inner2Static' {
+   method '<init> (Ljava/lang/String;)V' {
+      6      49
+      9      50
    }
 }
 
@@ -119,12 +119,12 @@ Lines mapping:
 39 <-> 27
 40 <-> 28
 41 <-> 29
-46 <-> 51
-47 <-> 52
-53 <-> 35
-54 <-> 36
-59 <-> 42
-60 <-> 43
+46 <-> 35
+47 <-> 36
+53 <-> 43
+54 <-> 44
+59 <-> 50
+60 <-> 51
 Not mapped:
 21
 33
diff --git a/testData/results/TestInnerLocalPkg.dec b/testData/results/TestInnerLocalPkg.dec
index 7cafa415f9b6212b593d0c0940cc16c277cf5e2f..7ddb0ce6ad8adf326a51e145f419e23eed8779dd 100644
--- a/testData/results/TestInnerLocalPkg.dec
+++ b/testData/results/TestInnerLocalPkg.dec
@@ -30,6 +30,14 @@ public class TestInnerLocalPkg {
       new TestInnerLocalPkg.Inner1Static.Inner2Static("test");// 42
    }// 43
 
+   class Inner1 {
+      final String x;
+
+      public Inner1(String var2) {
+         this.x = var2;// 48
+      }// 49
+   }
+
    static class Inner1Static {
       final String x;
 
@@ -45,14 +53,6 @@ public class TestInnerLocalPkg {
          }// 62
       }
    }
-
-   class Inner1 {
-      final String x;
-
-      public Inner1(String var2) {
-         this.x = var2;// 48
-      }// 49
-   }
 }
 
 class 'pkg/TestInnerLocalPkg$1Inner' {
@@ -86,24 +86,24 @@ class 'pkg/TestInnerLocalPkg$2Inner' {
    }
 }
 
-class 'pkg/TestInnerLocalPkg$Inner1Static' {
-   method '<init> (Ljava/lang/String;)V' {
-      6      36
-      9      37
+class 'pkg/TestInnerLocalPkg$Inner1' {
+   method '<init> (Lpkg/TestInnerLocalPkg;Ljava/lang/String;)V' {
+      b      36
+      e      37
    }
 }
 
-class 'pkg/TestInnerLocalPkg$Inner1Static$Inner2Static' {
+class 'pkg/TestInnerLocalPkg$Inner1Static' {
    method '<init> (Ljava/lang/String;)V' {
-      6      43
-      9      44
+      6      44
+      9      45
    }
 }
 
-class 'pkg/TestInnerLocalPkg$Inner1' {
-   method '<init> (Lpkg/TestInnerLocalPkg;Ljava/lang/String;)V' {
-      b      52
-      e      53
+class 'pkg/TestInnerLocalPkg$Inner1Static$Inner2Static' {
+   method '<init> (Ljava/lang/String;)V' {
+      6      51
+      9      52
    }
 }
 
@@ -121,12 +121,12 @@ Lines mapping:
 41 <-> 29
 42 <-> 30
 43 <-> 31
-48 <-> 53
-49 <-> 54
-55 <-> 37
-56 <-> 38
-61 <-> 44
-62 <-> 45
+48 <-> 37
+49 <-> 38
+55 <-> 45
+56 <-> 46
+61 <-> 52
+62 <-> 53
 Not mapped:
 23
 35
diff --git a/testData/results/TestInnerSignature.dec b/testData/results/TestInnerSignature.dec
index 3087a40dc63251ee1444ef4924050413f39221d5..406e1f41d8f0a084a986f683bf7ecb3b44c37429 100644
--- a/testData/results/TestInnerSignature.dec
+++ b/testData/results/TestInnerSignature.dec
@@ -9,18 +9,6 @@ public class TestInnerSignature<A, B, C> {
       this.c = var3;// 25
    }// 26
 
-   public static class InnerStatic<A, B, C> {
-      A a;
-      B b;
-      C c;
-
-      public InnerStatic(A var1, B var2, C var3) {
-         this.a = var1;// 46
-         this.b = var2;// 47
-         this.c = var3;// 48
-      }// 49
-   }
-
    public class Inner {
       A a;
       B b;
@@ -32,6 +20,18 @@ public class TestInnerSignature<A, B, C> {
          this.c = var4;// 36
       }// 37
    }
+
+   public static class InnerStatic<A, B, C> {
+      A a;
+      B b;
+      C c;
+
+      public InnerStatic(A var1, B var2, C var3) {
+         this.a = var1;// 46
+         this.b = var2;// 47
+         this.c = var3;// 48
+      }// 49
+   }
 }
 
 class 'TestInnerSignature' {
@@ -43,21 +43,21 @@ class 'TestInnerSignature' {
    }
 }
 
-class 'TestInnerSignature$InnerStatic' {
-   method '<init> (Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)V' {
-      6      17
-      b      18
-      10      19
-      13      20
+class 'TestInnerSignature$Inner' {
+   method '<init> (LTestInnerSignature;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)V' {
+      b      17
+      10      18
+      16      19
+      19      20
    }
 }
 
-class 'TestInnerSignature$Inner' {
-   method '<init> (LTestInnerSignature;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)V' {
-      b      29
-      10      30
-      16      31
-      19      32
+class 'TestInnerSignature$InnerStatic' {
+   method '<init> (Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)V' {
+      6      29
+      b      30
+      10      31
+      13      32
    }
 }
 
@@ -66,14 +66,14 @@ Lines mapping:
 24 <-> 8
 25 <-> 9
 26 <-> 10
-34 <-> 30
-35 <-> 31
-36 <-> 32
-37 <-> 33
-46 <-> 18
-47 <-> 19
-48 <-> 20
-49 <-> 21
+34 <-> 18
+35 <-> 19
+36 <-> 20
+37 <-> 21
+46 <-> 30
+47 <-> 31
+48 <-> 32
+49 <-> 33
 Not mapped:
 22
 33
diff --git a/testData/results/TestMethodParameters.dec b/testData/results/TestMethodParameters.dec
index 21f0f129d1bc260e2dbc7ce77f07ece57711bf75..af7ce865f5d946980010ee02c3e5be384a041064 100644
--- a/testData/results/TestMethodParameters.dec
+++ b/testData/results/TestMethodParameters.dec
@@ -21,6 +21,14 @@ public class TestMethodParameters {
 
    }// 39
 
+   class C1 {
+      C1(@Deprecated int var2) {
+      }// 24
+
+      void m(@Deprecated int var1) {
+      }// 25
+   }
+
    static class C2 {
       C2(@Deprecated int var1) {
       }// 29
@@ -31,14 +39,6 @@ public class TestMethodParameters {
       static void m2(@Deprecated int var0) {
       }// 31
    }
-
-   class C1 {
-      C1(@Deprecated int var2) {
-      }// 24
-
-      void m(@Deprecated int var1) {
-      }// 25
-   }
 }
 
 class 'pkg/TestMethodParameters' {
@@ -69,26 +69,26 @@ class 'pkg/TestMethodParameters$1Local' {
    }
 }
 
-class 'pkg/TestMethodParameters$C2' {
-   method '<init> (I)V' {
-      4      25
+class 'pkg/TestMethodParameters$C1' {
+   method '<init> (Lpkg/TestMethodParameters;I)V' {
+      9      25
    }
 
-   method 'm1 (I)V' {
+   method 'm (I)V' {
       0      28
    }
+}
 
-   method 'm2 (I)V' {
-      0      31
+class 'pkg/TestMethodParameters$C2' {
+   method '<init> (I)V' {
+      4      33
    }
-}
 
-class 'pkg/TestMethodParameters$C1' {
-   method '<init> (Lpkg/TestMethodParameters;I)V' {
-      9      36
+   method 'm1 (I)V' {
+      0      36
    }
 
-   method 'm (I)V' {
+   method 'm2 (I)V' {
       0      39
    }
 }
@@ -97,11 +97,11 @@ Lines mapping:
 19 <-> 5
 20 <-> 8
 21 <-> 11
-24 <-> 37
-25 <-> 40
-29 <-> 26
-30 <-> 29
-31 <-> 32
+24 <-> 26
+25 <-> 29
+29 <-> 34
+30 <-> 37
+31 <-> 40
 36 <-> 16
 37 <-> 19
 39 <-> 22
-- 
2.28.0.windows.1

