From 4e2882cb40ea460cb1f4a22c85a017cd853b308c Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Sun, 2 Aug 2015 21:14:56 -0700
Subject: Some synthetic anon classes are being detected as member classes.
 Unknown why, however this fixes the consturctor argument.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/NewExprent.java b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/NewExprent.java
index 6dad7c98c8c6c9fe0714e63c24763d2eb111df4c..0297a55fc16eb3eddc0e458bfa49325f563fd613 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/NewExprent.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/NewExprent.java
@@ -327,7 +327,7 @@ public class NewExprent extends Exprent {
 
                 if (i == lstParameters.size() - 1 && expr.getExprType() == VarType.VARTYPE_NULL) {
                   ClassNode node = DecompilerContext.getClassProcessor().getMapRootClasses().get(leftType.value);
-                  if (node != null && node.type == ClassNode.CLASS_ANONYMOUS) {
+                  if (node != null && (node.namelessConstructorStub || node.type == ClassNode.CLASS_ANONYMOUS)) {
                     break;  // skip last parameter of synthetic constructor call
                   }
                 }
-- 
2.28.0.windows.1

