From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Justin <jrd2558@gmail.com>
Date: Thu, 26 Jul 2018 13:28:40 -0700
Subject: [PATCH] Synthetic getClass/Objects.requireNonNull cleanup


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/ExprProcessor.java b/src/org/jetbrains/java/decompiler/modules/decompiler/ExprProcessor.java
index d5c4bf018352137c8491c3ac7628edb9106ddb56..c314bff6265a07deabd8d58bc6d02ff4f4c18e39 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/ExprProcessor.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/ExprProcessor.java
@@ -604,6 +604,23 @@ public class ExprProcessor implements CodeConstants {
           break;
         case opc_pop:
           stack.pop();
+          // check for synthetic getClass (J8) / Objects.requireNonNull() (J9+) calls added by the compiler
+          // see https://stackoverflow.com/a/20130641
+          if (i > 0) {
+            Exprent last = exprList.get(exprList.size() - 1);
+            if (last.type == Exprent.EXPRENT_ASSIGNMENT && ((AssignmentExprent)last).getRight().type == Exprent.EXPRENT_INVOCATION) {
+              InvocationExprent invocation = (InvocationExprent) ((AssignmentExprent) last).getRight();
+              if (i + 1 < seq.length()) {
+                if ((!invocation.isStatic() && invocation.getName().equals("getClass") && invocation.getStringDescriptor().equals("()Ljava/lang/Class;")) // J8
+                  || (invocation.isStatic() && invocation.getClassName().equals("java/util/Objects") && invocation.getName().equals("requireNonNull") && invocation.getStringDescriptor().equals("(Ljava/lang/Object;)Ljava/lang/Object;"))) { // J9+
+                  int nextOpc = seq.getInstr(i + 1).opcode;
+                  if (nextOpc >= opc_aconst_null && nextOpc <= opc_ldc2_w) {
+                    invocation.setSyntheticNullCheck();
+                  }
+                }
+              }
+            }
+          }
           break;
         case opc_pop2:
           if (stack.getByOffset(-1).getExprType().getStackSize() == 1) {
diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/SimplifyExprentsHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/SimplifyExprentsHelper.java
index e1bed2f0f5f501d05f93dfa6bfcd2f2b3e76b822..6d42db459055d505604728dcac4af7bb74a9e21c 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/SimplifyExprentsHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/SimplifyExprentsHelper.java
@@ -73,6 +73,10 @@ public class SimplifyExprentsHelper {
           if (changed) {
             break;
           }
+
+          if (!st.getStats().isEmpty() && hasQualifiedNewGetClass(st, st.getStats().get(0))) {
+            break;
+          }
         }
 
         res |= changed;
@@ -507,21 +511,53 @@ public class SimplifyExprentsHelper {
     return false;
   }
 
+  private static boolean hasQualifiedNewGetClass(Statement parent, Statement child) {
+    if (child.type == Statement.StatementType.BASIC_BLOCK && child.getExprents() != null && !child.getExprents().isEmpty()) {
+      Exprent firstExpr = child.getExprents().get(child.getExprents().size() - 1);
+
+      if (parent.type == Statement.StatementType.IF) {
+        if (isQualifiedNewGetClass(firstExpr, ((IfStatement)parent).getHeadexprent().getCondition())) {
+          child.getExprents().remove(firstExpr);
+          return true;
+        }
+      }
+      // TODO DoStatements ?
+    }
+    return false;
+  }
+
   private static boolean isQualifiedNewGetClass(Exprent first, Exprent second) {
     if (first.type == Exprent.EXPRENT_INVOCATION) {
       InvocationExprent invocation = (InvocationExprent)first;
 
-      if (!invocation.isStatic() && invocation.getInstance().type == Exprent.EXPRENT_VAR && invocation.getName().equals("getClass") &&
-          invocation.getStringDescriptor().equals("()Ljava/lang/Class;")) {
+      if ((!invocation.isStatic() &&
+           invocation.getName().equals("getClass") && invocation.getStringDescriptor().equals("()Ljava/lang/Class;")) // J8
+        || (invocation.isStatic() && invocation.getClassName().equals("java/util/Objects") && invocation.getName().equals("requireNonNull")
+            && invocation.getStringDescriptor().equals("(Ljava/lang/Object;)Ljava/lang/Object;"))) { // J9+
 
-        List<Exprent> lstExprents = second.getAllExprents();
+        if (invocation.isSyntheticNullCheck()) {
+          return true;
+        }
+
+        LinkedList<Exprent> lstExprents = new LinkedList<>();
         lstExprents.add(second);
 
-        for (Exprent expr : lstExprents) {
+        final Exprent target;
+        if (invocation.isStatic()) { // Objects.requireNonNull(target) (J9+)
+          // detect target type
+          target = invocation.getParameters().get(0);
+        } else { // target.getClass() (J8)
+          target = invocation.getInstance();
+        }
+
+        while (!lstExprents.isEmpty()) {
+          Exprent expr = lstExprents.removeFirst();
+          lstExprents.addAll(expr.getAllExprents());
           if (expr.type == Exprent.EXPRENT_NEW) {
             NewExprent newExpr = (NewExprent)expr;
             if (newExpr.getConstructor() != null && !newExpr.getConstructor().getParameters().isEmpty() &&
-                newExpr.getConstructor().getParameters().get(0).equals(invocation.getInstance())) {
+              (newExpr.getConstructor().getParameters().get(0).equals(target) ||
+                isUnambiguouslySameParam(invocation.isStatic(), target, newExpr.getConstructor().getParameters()))) {
 
               String classname = newExpr.getNewType().getValue();
               ClassNode node = DecompilerContext.getClassProcessor().getMapRootClasses().get(classname);
@@ -537,6 +573,27 @@ public class SimplifyExprentsHelper {
     return false;
   }
 
+  private static boolean isUnambiguouslySameParam(boolean isStatic, Exprent target, List<Exprent> parameters) {
+    boolean firstParamOfSameType = parameters.get(0).getExprType().equals(target.getExprType());
+    if (!isStatic) { // X.getClass()/J8, this is less likely to overlap with legitimate use
+      return firstParamOfSameType;
+    }
+    // Calling Objects.requireNonNull and discarding the result is a common pattern in normal code, so we have to be a bit more
+    // cautious about stripping calls when a constructor takes parameters of the instance type
+    // ex. given a class X, `Objects.requireNonNull(someInstanceOfX); new X(someInstanceOfX)` should not have the rNN stripped.
+    if (!firstParamOfSameType) {
+      return false;
+    }
+
+    for (int i = 1; i < parameters.size(); i++) {
+      if (parameters.get(i).getExprType().equals(target.getExprType())) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
   // propagate (var = new X) forward to the <init> invocation
   private static boolean isConstructorInvocationRemote(List<Exprent> list, int index) {
     Exprent current = list.get(index);
diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/InvocationExprent.java b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/InvocationExprent.java
index dbfe3abf4ec1f17d01771f514f662df1240faa11..d4c0672a5c8d1eb435d41c22d4fe22117ded5385 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/InvocationExprent.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/InvocationExprent.java
@@ -61,6 +61,7 @@ public class InvocationExprent extends Exprent {
   private List<VarType> genericArgs = new ArrayList<>();
   private boolean forceBoxing = false;
   private boolean forceUnboxing = false;
+  private boolean isSyntheticNullCheck = false;
 
   public InvocationExprent() {
     super(EXPRENT_INVOCATION);
@@ -163,6 +164,7 @@ public class InvocationExprent extends Exprent {
 
     addBytecodeOffsets(expr.bytecode);
     bootstrapArguments = expr.getBootstrapArguments();
+    isSyntheticNullCheck = expr.isSyntheticNullCheck();
   }
 
   @Override
@@ -965,6 +967,14 @@ public class InvocationExprent extends Exprent {
     return bootstrapArguments;
   }
 
+  public void setSyntheticNullCheck() {
+    isSyntheticNullCheck = true;
+  }
+
+  public boolean isSyntheticNullCheck() {
+    return isSyntheticNullCheck;
+  }
+
   @Override
   public void getBytecodeRange(BitSet values) {
     measureBytecode(values, parameters);
