From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Justin <jrd2558@gmail.com>
Date: Sat, 11 Jan 2020 17:00:41 -0700
Subject: [PATCH] Fix finally processor instruction comparison


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/FinallyProcessor.java b/src/org/jetbrains/java/decompiler/modules/decompiler/FinallyProcessor.java
index b240e83853cff82f917a7fc9f80dab661dd7ceb9..13317a543c5122c97c86801e63e0872a49df8f77 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/FinallyProcessor.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/FinallyProcessor.java
@@ -799,13 +799,17 @@ public class FinallyProcessor {
         int secondOp = second.operand(i);
         if (firstOp != secondOp) {
           // a-load/store instructions
-          if (first.opcode == CodeConstants.opc_aload || first.opcode == CodeConstants.opc_astore) {
+          if (first.opcode == CodeConstants.opc_aload) {
             for (int[] arr : lstStoreVars) {
               if (arr[0] == firstOp && arr[1] == secondOp) {
                 return true;
               }
             }
           }
+          else if (first.opcode == CodeConstants.opc_astore) {
+            lstStoreVars.add(new int[]{firstOp, secondOp});
+            return true;
+          }
 
           return false;
         }
