From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Justin <jrd2558@gmail.com>
Date: Fri, 20 Dec 2019 08:35:30 -0700
Subject: [PATCH] Simple lambda syntax support, --isl=0 to disable.


diff --git a/src/org/jetbrains/java/decompiler/main/ClassWriter.java b/src/org/jetbrains/java/decompiler/main/ClassWriter.java
index 6f2f8419d41a240c6d1428be0b0ca9258f058127..d9c338ff9189953329a0812250fb494656eb749a 100644
--- a/src/org/jetbrains/java/decompiler/main/ClassWriter.java
+++ b/src/org/jetbrains/java/decompiler/main/ClassWriter.java
@@ -10,14 +10,19 @@ import org.jetbrains.java.decompiler.main.rels.ClassWrapper;
 import org.jetbrains.java.decompiler.main.rels.MethodWrapper;
 import org.jetbrains.java.decompiler.modules.decompiler.ExprProcessor;
 import org.jetbrains.java.decompiler.modules.decompiler.exps.AnnotationExprent;
+import org.jetbrains.java.decompiler.modules.decompiler.exps.AssignmentExprent;
 import org.jetbrains.java.decompiler.modules.decompiler.exps.ConstExprent;
+import org.jetbrains.java.decompiler.modules.decompiler.exps.ExitExprent;
 import org.jetbrains.java.decompiler.modules.decompiler.exps.Exprent;
+import org.jetbrains.java.decompiler.modules.decompiler.exps.FunctionExprent;
 import org.jetbrains.java.decompiler.modules.decompiler.exps.NewExprent;
+import org.jetbrains.java.decompiler.modules.decompiler.exps.VarExprent;
 import org.jetbrains.java.decompiler.modules.decompiler.stats.RootStatement;
 import org.jetbrains.java.decompiler.modules.decompiler.typeann.TargetInfo;
 import org.jetbrains.java.decompiler.modules.decompiler.typeann.TypeAnnotation;
 import org.jetbrains.java.decompiler.modules.decompiler.typeann.TypeAnnotationWriteHelper;
 import org.jetbrains.java.decompiler.modules.decompiler.vars.VarProcessor;
+import org.jetbrains.java.decompiler.modules.decompiler.stats.Statement;
 import org.jetbrains.java.decompiler.modules.decompiler.vars.VarVersionPair;
 import org.jetbrains.java.decompiler.modules.renamer.PoolInterceptor;
 import org.jetbrains.java.decompiler.struct.*;
@@ -84,7 +89,15 @@ public class ClassWriter {
       if (node.lambdaInformation.is_method_reference) {
         if (!node.lambdaInformation.is_content_method_static && method_object != null) {
           // reference to a virtual method
-          buffer.append(method_object.toJava(indent, tracer));
+          method_object.getInferredExprType(new VarType(CodeConstants.TYPE_OBJECT, 0, node.lambdaInformation.content_class_name));
+          String instance = method_object.toJava(indent, tracer).toString();
+          // If the instance is casted, then we need to wrap it
+          if (method_object.type == Exprent.EXPRENT_FUNCTION && ((FunctionExprent)method_object).getFuncType() == FunctionExprent.FUNCTION_CAST && ((FunctionExprent)method_object).doesCast()) {
+            buffer.append('(').append(instance).append(')');
+          }
+          else {
+            buffer.append(instance);
+          }
         }
         else {
           // reference to a static method
@@ -103,6 +116,7 @@ public class ClassWriter {
 
         List<TypeAnnotation> parameterTypeAnnotations = TargetInfo.FormalParameterTarget.extract(TypeAnnotation.listFrom(mt));
         boolean explicitlyTyped = !parameterTypeAnnotations.isEmpty();
+        boolean simpleLambda = false;
 
         if (!lambdaToAnonymous) {
           buffer.append('(');
@@ -139,16 +153,60 @@ public class ClassWriter {
           }
 
           buffer.append(") ->");
+
+          RootStatement root = wrapper.getMethodWrapper(mt.getName(), mt.getDescriptor()).root;
+          if (DecompilerContext.getOption(IFernflowerPreferences.INLINE_SIMPLE_LAMBDAS) && !methodWrapper.decompiledWithErrors && root != null) {
+            Statement firstStat = root.getFirst();
+            if (firstStat.type == Statement.StatementType.BASIC_BLOCK && firstStat.getExprents() != null && firstStat.getExprents().size() == 1) {
+              Exprent firstExpr = firstStat.getExprents().get(0);
+              boolean isVarDefinition = firstExpr.type == Exprent.EXPRENT_ASSIGNMENT &&
+                ((AssignmentExprent)firstExpr).getLeft().type == Exprent.EXPRENT_VAR &&
+                ((VarExprent)((AssignmentExprent)firstExpr).getLeft()).isDefinition();
+
+              boolean isThrow = firstExpr.type == Exprent.EXPRENT_EXIT &&
+                ((ExitExprent)firstExpr).getExitType() == ExitExprent.EXIT_THROW;
+
+              if (!isVarDefinition && !isThrow) {
+                simpleLambda = true;
+                MethodWrapper outerWrapper = (MethodWrapper)DecompilerContext.getProperty(DecompilerContext.CURRENT_METHOD_WRAPPER);
+                DecompilerContext.setProperty(DecompilerContext.CURRENT_METHOD_WRAPPER, methodWrapper);
+                try {
+                  TextBuffer codeBuffer = firstExpr.toJava(indent + 1, tracer);
+
+                  if (firstExpr.type == Exprent.EXPRENT_EXIT)
+                    codeBuffer.setStart(6); // skip return
+                  else
+                    codeBuffer.prepend(" ");
+
+                  buffer.append(codeBuffer);
+                }
+                catch (Throwable ex) {
+                  DecompilerContext.getLogger().writeMessage("Method " + mt.getName() + " " + mt.getDescriptor() + " couldn't be written.",
+                    IFernflowerLogger.Severity.WARN,
+                    ex);
+                  methodWrapper.decompiledWithErrors = true;
+                  buffer.append(" // $FF: Couldn't be decompiled");
+                }
+                finally {
+                  tracer.addMapping(root.getDummyExit().bytecode);
+                  addTracer(cl, mt, tracer);
+                  DecompilerContext.setProperty(DecompilerContext.CURRENT_METHOD_WRAPPER, outerWrapper);
+                }
+              }
+            }
+          }
         }
 
-        buffer.append(" {").appendLineSeparator();
-        tracer.incrementCurrentSourceLine();
+        if (!simpleLambda) {
+          buffer.append(" {").appendLineSeparator();
+          tracer.incrementCurrentSourceLine();
 
-        methodLambdaToJava(node, wrapper, mt, buffer, indent + 1, !lambdaToAnonymous, tracer);
+          methodLambdaToJava(node, wrapper, mt, buffer, indent + 1, !lambdaToAnonymous, tracer);
 
-        buffer.appendIndent(indent).append("}");
+          buffer.appendIndent(indent).append("}");
 
-        addTracer(cl, mt, tracer);
+          addTracer(cl, mt, tracer);
+        }
       }
     }
     finally {
diff --git a/src/org/jetbrains/java/decompiler/main/extern/IFernflowerPreferences.java b/src/org/jetbrains/java/decompiler/main/extern/IFernflowerPreferences.java
index a25566f0e6ba9a4f67c6412fc379874ffca92c21..f5e95e65a807e335ce33a589348e11ad1dbbf5fe 100644
--- a/src/org/jetbrains/java/decompiler/main/extern/IFernflowerPreferences.java
+++ b/src/org/jetbrains/java/decompiler/main/extern/IFernflowerPreferences.java
@@ -39,6 +39,7 @@ public interface IFernflowerPreferences {
   String INCLUDE_ENTIRE_CLASSPATH = "iec";
   String QUALIFY_INNER_CLASSES = "qin";
   String EXPLICIT_GENERIC_ARGUMENTS = "ega";
+  String INLINE_SIMPLE_LAMBDAS = "isl";
 
   String LOG_LEVEL = "log";
   String MAX_PROCESSING_METHOD = "mpm";
@@ -95,6 +96,7 @@ public interface IFernflowerPreferences {
     defaults.put(INCLUDE_ENTIRE_CLASSPATH, "0");
     defaults.put(QUALIFY_INNER_CLASSES, "1");
     defaults.put(EXPLICIT_GENERIC_ARGUMENTS, "0");
+    defaults.put(INLINE_SIMPLE_LAMBDAS, "1");
 
     defaults.put(LOG_LEVEL, IFernflowerLogger.Severity.INFO.name());
     defaults.put(MAX_PROCESSING_METHOD, "0");
diff --git a/test/org/jetbrains/java/decompiler/SingleClassesTest.java b/test/org/jetbrains/java/decompiler/SingleClassesTest.java
index 16ceb1e8e47c90da72ef3039cf625eec25b68223..c78c9554df9788b87028bcf181fc243db7a29f3f 100644
--- a/test/org/jetbrains/java/decompiler/SingleClassesTest.java
+++ b/test/org/jetbrains/java/decompiler/SingleClassesTest.java
@@ -24,7 +24,8 @@ public class SingleClassesTest extends SingleClassesTestBase {
                          IFernflowerPreferences.DUMP_ORIGINAL_LINES, "1",
                          IFernflowerPreferences.IGNORE_INVALID_BYTECODE, "1",
                          IFernflowerPreferences.VERIFY_ANONYMOUS_CLASSES, "1",
-                         IFernflowerPreferences.INCLUDE_ENTIRE_CLASSPATH, "1"
+                         IFernflowerPreferences.INCLUDE_ENTIRE_CLASSPATH, "1",
+                         IFernflowerPreferences.INLINE_SIMPLE_LAMBDAS, "0"
 );
   }
 
