From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Fri, 2 Oct 2015 03:10:28 -0700
Subject: [PATCH] Include LVT info in VarExprent.copy


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/VarExprent.java b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/VarExprent.java
index 933b0467478a4315a6d8dbdb7ae067f57046a8bc..c85d889d96073327a5ea08a9d986621dd8792e26 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/VarExprent.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/VarExprent.java
@@ -82,6 +82,7 @@ public class VarExprent extends Exprent {
     var.setVersion(version);
     var.setClassDef(classDef);
     var.setStack(stack);
+    var.setLVT(lvt);
     return var;
   }
 
diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
index 2259ce1c9b2bc515fd588ff44a2d152567ab3cf4..bec695f0265d7ed92175ba4e43ce2932f70eed16 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
@@ -260,22 +260,18 @@ public class VarDefinitionHelper {
   }
 
   private LVTVariable findLVT(int index, Exprent exp) {
-    VarExprent var = null;
-
-    if (exp.type == Exprent.EXPRENT_ASSIGNMENT) {
-      AssignmentExprent ass = (AssignmentExprent)exp;
-      if (ass.getLeft().type == Exprent.EXPRENT_VAR) {
-        var = (VarExprent)ass.getLeft();
+    for (Exprent e : exp.getAllExprents(false)) {
+      LVTVariable lvt = findLVT(index, e);
+      if (lvt != null) {
+        return lvt;
       }
     }
-    else if (exp.type == Exprent.EXPRENT_VAR) {
-      var = (VarExprent)exp;
-    }
 
-    if (var == null) {
+    if (exp.type != Exprent.EXPRENT_VAR) {
       return null;
     }
 
+    VarExprent var = (VarExprent)exp;
     return var.getIndex() == index ? var.getLVT() : null;
   }
 
diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarProcessor.java b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarProcessor.java
index bbf17adc460d22b1c91fd54ba88d0e2624c625c4..89f24f365a7c01b7920fd8002dda2ec0b955f96c 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarProcessor.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarProcessor.java
@@ -166,7 +166,7 @@ public class VarProcessor {
   public void findLVT(VarExprent varExprent, int bytecodeOffset) {
     LVTVariable var = this.lvt == null ? null : lvt.find(varExprent.getIndex(), bytecodeOffset);
     if (var != null) {
-        varExprent.setLVT(var);
+      varExprent.setLVT(var);
     }
   }
 
