From 86199e1f573a00ff9095aca30e0f3eb422a0d9a8 Mon Sep 17 00:00:00 2001
From: Justin <jrd2558@gmail.com>
Date: Sat, 11 Jan 2020 17:00:41 -0700
Subject: [PATCH] Fix finally processor instruction comparison


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/FinallyProcessor.java b/src/org/jetbrains/java/decompiler/modules/decompiler/FinallyProcessor.java
index 59b014e..d68fb07 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/FinallyProcessor.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/FinallyProcessor.java
@@ -856,13 +856,17 @@ public class FinallyProcessor {
         int secondOp = second.operand(i);
         if (firstOp != secondOp) {
           // a-load/store instructions
-          if (first.opcode == CodeConstants.opc_aload || first.opcode == CodeConstants.opc_astore) {
+          if (first.opcode == CodeConstants.opc_aload) {
             for (int[] arr : lstStoreVars) {
               if (arr[0] == firstOp && arr[1] == secondOp) {
                 return true;
               }
             }
           }
+          else if (first.opcode == CodeConstants.opc_astore) {
+            lstStoreVars.add(new int[]{firstOp, secondOp});
+            return true;
+          }
 
           return false;
         }
-- 
2.17.2 (Apple Git-113)

