From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LexManos <LexManos@gmail.com>
Date: Mon, 15 Jun 2020 22:07:51 -0700
Subject: [PATCH] Fix local variables incorrectly merging.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/StackVarsProcessor.java b/src/org/jetbrains/java/decompiler/modules/decompiler/StackVarsProcessor.java
index 139ccdb55347a5d66627c1489ee73910fed46f05..388ead266fe10aa4c251d3b7c6207e20126e66b9 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/StackVarsProcessor.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/StackVarsProcessor.java
@@ -602,7 +602,7 @@ public class StackVarsProcessor {
 
     for (Entry<Integer, Set<VarVersionPair>> ent : mapVars.entrySet()) {
       FastSparseSet<Integer> liveverset = mapLiveVars.get(ent.getKey());
-      if (liveverset == null) {
+      if (liveverset == null || liveverset.isEmpty()) {
         return false;
       }
 
@@ -611,13 +611,13 @@ public class StackVarsProcessor {
         domset.add(ssuversions.nodes.getWithKey(verpaar));
       }
 
-      boolean isdom = false;
+      boolean isdom = true;
 
       for (Integer livever : liveverset) {
         VarVersionNode node = ssuversions.nodes.getWithKey(new VarVersionPair(ent.getKey().intValue(), livever.intValue()));
 
-        if (ssuversions.isDominatorSet(node, domset)) {
-          isdom = true;
+        if (!ssuversions.isDominatorSet(node, domset)) {
+          isdom = false;
           break;
         }
       }
