From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LexManos <LexManos@gmail.com>
Date: Wed, 17 Jun 2020 11:08:27 -0700
Subject: [PATCH] Add toString to MethodDescriptor


diff --git a/src/org/jetbrains/java/decompiler/struct/gen/MethodDescriptor.java b/src/org/jetbrains/java/decompiler/struct/gen/MethodDescriptor.java
index 2096754661e845cfe3148d47c5c3ff914bcb0322..c033ae3122302a16f760df24110ab3dd07e39852 100644
--- a/src/org/jetbrains/java/decompiler/struct/gen/MethodDescriptor.java
+++ b/src/org/jetbrains/java/decompiler/struct/gen/MethodDescriptor.java
@@ -19,11 +19,13 @@ import java.util.Objects;
 public final class MethodDescriptor {
   public final VarType[] params;
   public final VarType ret;
+  private final String descriptor;
   public GenericMethodDescriptor genericInfo;
 
-  private MethodDescriptor(VarType[] params, VarType ret) {
+  private MethodDescriptor(VarType[] params, VarType ret, String descriptor) {
     this.params = params;
     this.ret = ret;
+    this.descriptor = descriptor;
   }
 
   public static MethodDescriptor parseDescriptor(String descriptor) {
@@ -71,7 +73,7 @@ public final class MethodDescriptor {
 
     VarType ret = new VarType(descriptor.substring(parenth + 1));
 
-    return new MethodDescriptor(params, ret);
+    return new MethodDescriptor(params, ret, descriptor);
   }
 
   public static MethodDescriptor parseDescriptor(StructMethod struct, ClassNode node) {
@@ -150,6 +152,11 @@ public final class MethodDescriptor {
     return null;
   }
 
+  @Override
+  public String toString() {
+    return this.descriptor;
+  }
+
   @Override
   public boolean equals(Object o) {
     if (o == this) return true;
