From 7b1e12d9f6b659bf671250d3e24b1314f0129428 Mon Sep 17 00:00:00 2001
From: LexManos <LexManos@gmail.com>
Date: Wed, 17 Jun 2020 11:08:27 -0700
Subject: [PATCH] Add toString to MethodDescriptor


diff --git a/src/org/jetbrains/java/decompiler/struct/gen/MethodDescriptor.java b/src/org/jetbrains/java/decompiler/struct/gen/MethodDescriptor.java
index 49bb8c8..4995d41 100644
--- a/src/org/jetbrains/java/decompiler/struct/gen/MethodDescriptor.java
+++ b/src/org/jetbrains/java/decompiler/struct/gen/MethodDescriptor.java
@@ -19,11 +19,13 @@ import java.util.Objects;
 public class MethodDescriptor {
   public final VarType[] params;
   public final VarType ret;
+  private final String descriptor;
   public GenericMethodDescriptor genericInfo;
 
-  private MethodDescriptor(VarType[] params, VarType ret) {
+  private MethodDescriptor(VarType[] params, VarType ret, String descriptor) {
     this.params = params;
     this.ret = ret;
+    this.descriptor = descriptor;
   }
 
   public static MethodDescriptor parseDescriptor(String descriptor) {
@@ -70,7 +72,7 @@ public class MethodDescriptor {
 
     VarType ret = new VarType(descriptor.substring(parenth + 1));
 
-    return new MethodDescriptor(params, ret);
+    return new MethodDescriptor(params, ret, descriptor);
   }
 
   public static MethodDescriptor parseDescriptor(StructMethod struct, ClassNode node) {
@@ -152,6 +154,11 @@ public class MethodDescriptor {
     return null;
   }
 
+  @Override
+  public String toString() {
+    return this.descriptor;
+  }
+
   @Override
   public boolean equals(Object o) {
     if (o == this) return true;
-- 
2.27.0.windows.1

