From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cpw <cpw@weeksfamily.ca>
Date: Thu, 27 Aug 2015 15:41:47 -0400
Subject: [PATCH] Fix inverted null test


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
index 1e1a4a7987545d00460cd7e7e17c3bcde6a75df1..d24813f0d9962ed8b0d6efacd251c175721e33d0 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
@@ -556,7 +556,7 @@ public class VarDefinitionHelper {
   }
 
   private void setupLVTs(Statement stat) {
-    if (stat == null || varproc.getLVT() != null) {
+    if (stat == null || varproc.getLVT() == null) {
       return;
     }
 
