From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: coehlrich <coehlrich@users.noreply.github.com>
Date: Tue, 16 Nov 2021 18:22:31 +1300
Subject: [PATCH] Remove default keyword from private interface methods


diff --git a/src/org/jetbrains/java/decompiler/main/ClassWriter.java b/src/org/jetbrains/java/decompiler/main/ClassWriter.java
index 18e042dfff071da0efa578dc3966bc93ad116d47..05deb2b78d448fee16329617571a0e07e13d8570 100644
--- a/src/org/jetbrains/java/decompiler/main/ClassWriter.java
+++ b/src/org/jetbrains/java/decompiler/main/ClassWriter.java
@@ -829,7 +829,7 @@ public class ClassWriter {
 
       appendModifiers(buffer, flags, METHOD_ALLOWED, isInterface, METHOD_EXCLUDED);
 
-      if (isInterface && !mt.hasModifier(CodeConstants.ACC_STATIC) && mt.containsCode()) {
+      if (isInterface && !mt.hasModifier(CodeConstants.ACC_STATIC) && !mt.hasModifier(CodeConstants.ACC_PRIVATE) && mt.containsCode()) {
         // 'default' modifier (Java 8)
         buffer.append("default ");
       }
