From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cpw <cpw@weeksfamily.ca>
Date: Fri, 16 Oct 2015 21:39:10 -0400
Subject: [PATCH] Tweak decompilation test to new settings (no synthetics or
 bridge)


diff --git a/test/org/jetbrains/java/decompiler/MinecraftDecompilationTest.java b/test/org/jetbrains/java/decompiler/MinecraftDecompilationTest.java
index 2afd2e6f60ee5cc191b6eaf535e0fa12f25ea275..087c8102c11b1258d8e3f1615fe64f745b8587d6 100644
--- a/test/org/jetbrains/java/decompiler/MinecraftDecompilationTest.java
+++ b/test/org/jetbrains/java/decompiler/MinecraftDecompilationTest.java
@@ -43,10 +43,8 @@ public class MinecraftDecompilationTest {
     // -din=1 -rbr=0 -dgs=1 -asc=1 -rsy=0
     Map<String,Object> mcFFOptions = new HashMap<String,Object>() {{
         put(IFernflowerPreferences.DECOMPILE_INNER,"1");
-        put(IFernflowerPreferences.REMOVE_BRIDGE, "0");
         put(IFernflowerPreferences.DECOMPILE_GENERIC_SIGNATURES,"1");
         put(IFernflowerPreferences.ASCII_STRING_CHARACTERS,"1");
-        put(IFernflowerPreferences.REMOVE_SYNTHETIC,"0");
         put(IFernflowerPreferences.INCLUDE_ENTIRE_CLASSPATH, "1");
     }};
     fixture.setUp(mcFFOptions);
