From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: coehlrich <coehlrich@users.noreply.github.com>
Date: Sun, 27 Mar 2022 14:22:14 +1300
Subject: [PATCH] Fix sealed enums


diff --git a/src/org/jetbrains/java/decompiler/main/ClassWriter.java b/src/org/jetbrains/java/decompiler/main/ClassWriter.java
index a33744bb55dc0fa01514b4866ccef14ecdc3e5d5..acfa1ad5ffc129d478423ed567efb9dc3c046931 100644
--- a/src/org/jetbrains/java/decompiler/main/ClassWriter.java
+++ b/src/org/jetbrains/java/decompiler/main/ClassWriter.java
@@ -415,7 +415,7 @@ public class ClassWriter {
 
     appendModifiers(buffer, flags, CLASS_ALLOWED, isInterface, CLASS_EXCLUDED);
 
-    if (permittedSubclassQualifiedNames != null) {
+    if (permittedSubclassQualifiedNames != null && !isEnum) {
       buffer.append("sealed ");
     }
     else if (node.isNonSealed()) {
