From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Sun, 13 Sep 2015 18:19:20 -0700
Subject: [PATCH] Prevent infinite loops in PPMM.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/PPandMMHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/PPandMMHelper.java
index f39946d1c839333a69c91521f6714b183028567d..db70f64a91a21c04b1277b6da9b38ef7f60c1cf4 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/PPandMMHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/PPandMMHelper.java
@@ -157,7 +157,11 @@ public class PPandMMHelper {
 
               exprentReplaced = true;
               if (!left.equals(econd)) {
-                remaps.put(new VarVersionPair((VarExprent)left), new VarVersionPair((VarExprent)econd));
+                VarVersionPair _old = new VarVersionPair((VarExprent)left);
+                VarVersionPair _new = new VarVersionPair((VarExprent)econd);
+                if (_old.var > _new.var) { // Try and prevent infinite loops here by only merging 'down'
+                  remaps.put(_old, _new);
+                }
               }
               return ret;
             }
