From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cpw <cpw@weeksfamily.ca>
Date: Thu, 24 Sep 2015 00:00:19 -0400
Subject: [PATCH] If it's a null constant, don't give it a type. Cleans up some
 more errors.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
index 0c03cd52d3ab4b664545ebf8c57faa9381a7ac65..9ecb28bffa6dc6159370823066d8f189a3f1a83b 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
@@ -637,6 +637,9 @@ if (mt.getName().equals("func_180655_c")){
           }
 
           ConstExprent right = (ConstExprent)ass.getRight();
+          if (right.getConstType() == VarType.VARTYPE_NULL) {
+              continue;
+          }
           VarType merged = getMergedType(mapExprentMinTypes.get(from), mapExprentMinTypes.get(to),
                                          mapExprentMaxTypes.get(from), mapExprentMaxTypes.get(to));
 
diff --git a/test/org/jetbrains/java/decompiler/LVTTest.java b/test/org/jetbrains/java/decompiler/LVTTest.java
index d2a8f1c20c8a7954d4ba22d17fc1584d3bd4a3ca..52428c70df9fe4166129c11eacd4d19c2d7196e3 100644
--- a/test/org/jetbrains/java/decompiler/LVTTest.java
+++ b/test/org/jetbrains/java/decompiler/LVTTest.java
@@ -35,10 +35,20 @@ public class LVTTest extends SingleClassesTestBase {
         super.setUp();
         fixture.cleanup = false;
     }
-  @Test public void testMatch1() { doTest("pkg/TestPPMM"); }
+//  @Test public void testMatch1() { doTest("pkg/TestPPMM"); }
 //  @Test public void testMatchLM() { doTest("pkg/TestLexManosLVT"); }
 //  @Test public void testMatch1() { doTest("pkg/TestLVT"); }
 //  @Test public void testMatch2() { doTest("pkg/TestLVTScoping"); }
 //  @Test public void testMCWorld() { doTest("net/minecraft/world/World"); }
 //  @Test public void testMCGuiCmdBlock() { doTest("net/minecraft/client/gui/GuiCommandBlock"); }
+
+//  @Test public void testMCWorld() { doTest("net/minecraft/world/World"); }
+
+//  @Test public void testMCGuiCmdBlock() { doTest("net/minecraft/client/gui/GuiCommandBlock"); }
+
+//  @Test public void testMCBlockFence() { doTest("net/minecraft/block/BlockFence"); }
+
+//  @Test public void testMCAbstractTexture() { doTest("net/minecraft/client/multiplayer/ServerAddress"); }
+//  @Test public void testMCAbstractTexture() { doTest("net/minecraft/client/resources/AbstractResourcePack"); }
+  @Test public void testMCGuiShareToLan() { doTest("net/minecraft/client/gui/GuiShareToLan"); }
 }
