From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Thu, 1 Oct 2015 18:03:36 -0700
Subject: [PATCH] NPE protection in LVT loading.


diff --git a/src/org/jetbrains/java/decompiler/struct/attr/StructLocalVariableTableAttribute.java b/src/org/jetbrains/java/decompiler/struct/attr/StructLocalVariableTableAttribute.java
index 771ec7add154d7897fa54ac46d91e66e6ad2ecc5..a95870f4b9de4b87eff9a15b88492e8cc051cbdd 100644
--- a/src/org/jetbrains/java/decompiler/struct/attr/StructLocalVariableTableAttribute.java
+++ b/src/org/jetbrains/java/decompiler/struct/attr/StructLocalVariableTableAttribute.java
@@ -61,8 +61,13 @@ public class StructLocalVariableTableAttribute extends StructGeneralAttribute {
   }
 
   public void addLocalVariableTable(StructLocalVariableTableAttribute attr) {
-    lvt.mergeLVTs(attr.lvt);
-    attr.lvt = lvt;
+    if (lvt == null) {
+      lvt = attr.lvt;
+    }
+    else {
+      lvt.mergeLVTs(attr.lvt);
+      attr.lvt = lvt;
+    }
   }
 
   public Map<Integer, List<LVTVariable>> getMapVarNames() {
