From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Thu, 1 Oct 2015 18:04:09 -0700
Subject: [PATCH] Fix variable defintions in IfStatement's headers having wrong
 scope.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
index cc5e5e427125ab4673f9aef2a7b56de37f0ab113..560328a405e5860dba08daf85f0e957a21fc91d1 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
@@ -531,7 +531,7 @@ public class VarDefinitionHelper {
           var.setLVT(lvt);
         } else {
           System.currentTimeMillis();
-          System.out.println("null " + new VarVersionPair(var));
+          //System.out.println("null " + new VarVersionPair(var));
         }
       }
     }
@@ -631,6 +631,9 @@ public class VarDefinitionHelper {
                 leaked_n.clear(); // Force no leaking at the end of if blocks
                 // We may need to do this for Switches as well.. But havent run into that issue yet...
               }
+              else if (obj == ifst.getFirst()) {
+                leaked.putAll(leaked_n); //First is outside the scope so leak!
+              }
             }
             this_vars.putAll(leaked_n);
           }
