From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Fri, 2 Oct 2015 01:32:57 -0700
Subject: [PATCH] NPE Protection in EnumProcessor


diff --git a/src/org/jetbrains/java/decompiler/main/EnumProcessor.java b/src/org/jetbrains/java/decompiler/main/EnumProcessor.java
index 37a3384c390209bdbcf46843a4c1a771a24dde18..8956e0ebbe08c761ab14ed61850d767b46642a99 100644
--- a/src/org/jetbrains/java/decompiler/main/EnumProcessor.java
+++ b/src/org/jetbrains/java/decompiler/main/EnumProcessor.java
@@ -74,7 +74,9 @@ public class EnumProcessor {
 
   // FIXME: move to a util class (see also InitializerProcessor)
   private static Statement findFirstData(Statement stat) {
-
+    if (stat == null) {
+      return null;
+    }
     if (stat.getExprents() != null) {
       return stat;
     }
