From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Sun, 4 Oct 2015 02:10:44 -0700
Subject: [PATCH] Hack to prevent extracted ifs from being inlined again in
 infinite loop. MC decompiles much better!


diff --git a/src/org/jetbrains/java/decompiler/main/rels/MethodProcessorRunnable.java b/src/org/jetbrains/java/decompiler/main/rels/MethodProcessorRunnable.java
index eca34134bf72cb0ddc9e3b245ca3847478f3a39f..62d52975708c5a8126bdc73824fe575b14dea108 100644
--- a/src/org/jetbrains/java/decompiler/main/rels/MethodProcessorRunnable.java
+++ b/src/org/jetbrains/java/decompiler/main/rels/MethodProcessorRunnable.java
@@ -33,14 +33,12 @@ import org.jetbrains.java.decompiler.modules.decompiler.exps.VarExprent;
 import org.jetbrains.java.decompiler.modules.decompiler.stats.DummyExitStatement;
 import org.jetbrains.java.decompiler.modules.decompiler.stats.RootStatement;
 import org.jetbrains.java.decompiler.modules.decompiler.stats.Statement;
-import org.jetbrains.java.decompiler.modules.decompiler.vars.LVTVariable;
 import org.jetbrains.java.decompiler.modules.decompiler.vars.VarProcessor;
 import org.jetbrains.java.decompiler.struct.StructClass;
 import org.jetbrains.java.decompiler.struct.StructMethod;
 
 import java.io.IOException;
 import java.util.BitSet;
-import java.util.List;
 
 public class MethodProcessorRunnable implements Runnable {
 
diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/ExitHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/ExitHelper.java
index d467cdcc200f5b0547a94d7276e00dc8521f31fa..df2cb4a554931467d7feccaf1c0393192eeb3bc2 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/ExitHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/ExitHelper.java
@@ -206,7 +206,7 @@ public class ExitHelper {
     Statement dest = edge.getDestination();
 
     if (edge.getType() == StatEdge.TYPE_BREAK && dest.type == Statement.TYPE_BASICBLOCK
-        && edge.explicit && (edge.labeled || isOnlyEdge(edge))) {
+        && edge.explicit && (edge.labeled || isOnlyEdge(edge)) && edge.canInline) {
       List<Exprent> data = dest.getExprents();
 
       if (data != null && data.size() == 1) {
diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/InlineSingleBlockHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/InlineSingleBlockHelper.java
index adeda024151b030fe94e7ab4b224dc51dc6e8ea8..1c468c818be14ab4a92392b48db98f0ff72bb2b2 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/InlineSingleBlockHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/InlineSingleBlockHelper.java
@@ -132,7 +132,10 @@ public class InlineSingleBlockHelper {
       StatEdge edge = lst.get(0);
 
       if (sameCatchRanges(edge)) {
-        if (edge.explicit) {
+        if (!edge.canInline) {
+          return false; //Dirty hack, but lets do it!
+        }
+        else if (edge.explicit) {
           return true;
         }
         else {
diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/LoopExtractHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/LoopExtractHelper.java
index 93bc21dd17add5a854450a034ed69a0a961808ee..f09dd3c0a4251b0a7c792fff20635704f4c8784a 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/LoopExtractHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/LoopExtractHelper.java
@@ -92,8 +92,14 @@ public class LoopExtractHelper {
       }
     }
 
-    if (!extractLastIf(stat, stats)) {
-      return extractFirstIf(stat, stats);
+    if (stats.size() > 0) { // In this case prioritize first to help the Loop enhancer.
+      if (stat.getParent().getStats().getLast() != stat){
+        return false;
+      }
+    }
+
+    if (!extractFirstIf(stat, stats)) {
+      return extractLastIf(stat, stats);
     }
     else {
       return true;
@@ -219,5 +225,10 @@ public class LoopExtractHelper {
         loop.addPredecessor(edge);
       }
     }
+
+    List<StatEdge> link = target.getPredecessorEdges(StatEdge.TYPE_BREAK);
+    if (link.size() == 1) {
+      link.get(0).canInline = false;
+    }
   }
 }
diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/StatEdge.java b/src/org/jetbrains/java/decompiler/modules/decompiler/StatEdge.java
index 26d427b9377097653912b624ad76bd78563d5718..7902069a9fe823f4873c2fb00b0f4b308dfc454a 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/StatEdge.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/StatEdge.java
@@ -52,6 +52,8 @@ public class StatEdge {
 
   public boolean explicit = true;
 
+  public boolean canInline = true;
+
   public StatEdge(int type, Statement source, Statement destination, Statement closure) {
     this(type, source, destination);
     this.closure = closure;
diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/stats/SequenceStatement.java b/src/org/jetbrains/java/decompiler/modules/decompiler/stats/SequenceStatement.java
index 81d31a58f6a913ddc4f5a89ae22fe69f98c63f3e..7d2ee435ce1c153b6da410d825d697d9790e35fe 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/stats/SequenceStatement.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/stats/SequenceStatement.java
@@ -20,7 +20,6 @@ import org.jetbrains.java.decompiler.main.collectors.BytecodeMappingTracer;
 import org.jetbrains.java.decompiler.modules.decompiler.DecHelper;
 import org.jetbrains.java.decompiler.modules.decompiler.ExprProcessor;
 import org.jetbrains.java.decompiler.modules.decompiler.StatEdge;
-import org.jetbrains.java.decompiler.modules.decompiler.vars.StartEndPair;
 
 import java.util.Arrays;
 import java.util.List;
