From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Sun, 4 Oct 2015 04:22:34 -0700
Subject: [PATCH] Include ForEach source vars in itteration, allows for
 remapping in merge.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/stats/DoStatement.java b/src/org/jetbrains/java/decompiler/modules/decompiler/stats/DoStatement.java
index 36cd7fb229ea03784017336da4dd1cb0fd528fa4..acc74a60fd6a90ad15959f6c3164757390542ada 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/stats/DoStatement.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/stats/DoStatement.java
@@ -164,6 +164,7 @@ public class DoStatement extends Statement {
         break;
       case LOOP_FOREACH:
         lst.add(getInitExprent());
+        lst.add(getIncExprent());
     }
 
     lst.add(first);
