From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Sun, 4 Oct 2015 18:14:24 -0700
Subject: [PATCH] Fix issue where For loops would be enhanced to ForEach
 incorrectly.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/MergeHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/MergeHelper.java
index f354f0684b4d16e8dcd08b2ce6db76210249b571..11c7c599d4413cb46b58906f4b7fdbe4a019e1f0 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/MergeHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/MergeHelper.java
@@ -501,10 +501,12 @@ public class MergeHelper {
         }
 
         if (initExprents[0].getRight().type != Exprent.EXPRENT_CONST ||
-            initExprents[1].getRight().type != Exprent.EXPRENT_FUNCTION) {
+            initExprents[1].getRight().type != Exprent.EXPRENT_FUNCTION ||
+            stat.getConditionExprent().type != Exprent.EXPRENT_FUNCTION) {
           return false;
         }
 
+        //FunctionExprent funcCond  = (FunctionExprent)drillNots(stat.getConditionExprent()); //TODO: Verify this is counter < copy.length
         FunctionExprent funcRight = (FunctionExprent)initExprents[1].getRight();
         FunctionExprent funcInc   = (FunctionExprent)lastExprent;
         ArrayExprent    arr       = (ArrayExprent)firstDoExprent.getRight();
@@ -525,6 +527,9 @@ public class MergeHelper {
             counter.getVersion() != index.getVersion()) {
           return false;
         }
+        if (ExprentUtil.isVarReferenced(counter, stat.getFirst(), index)) {
+          return false;
+        }
 
         funcRight.getLstOperands().get(0).addBytecodeOffsets(initExprents[0].bytecode);
         funcRight.getLstOperands().get(0).addBytecodeOffsets(initExprents[1].bytecode);
