From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Wed, 14 Oct 2015 13:20:13 -0700
Subject: [PATCH] Fixed generic type parsing of inner classes with generic
 outers.


diff --git a/src/org/jetbrains/java/decompiler/struct/gen/generics/GenericType.java b/src/org/jetbrains/java/decompiler/struct/gen/generics/GenericType.java
index c45d47f7e7efbd28e5bbddb02441299306518733..aeb786a7b26864424c5b730ef3f954b9bfb583be 100644
--- a/src/org/jetbrains/java/decompiler/struct/gen/generics/GenericType.java
+++ b/src/org/jetbrains/java/decompiler/struct/gen/generics/GenericType.java
@@ -95,7 +95,7 @@ public class GenericType extends VarType {
               }
 
               if (name_buff.length() > 0) {
-                name_buff.append('.');
+                name_buff.append('$');
               }
               name_buff.append(name);
 
@@ -115,6 +115,7 @@ public class GenericType extends VarType {
 
                 signature = signature.substring(cl.length() + 1);
               }
+              cl = getNextClassSignature(signature);
             }
           }
           break loop;
