From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Wed, 14 Oct 2015 16:07:14 -0700
Subject: [PATCH] Itterator enchanced for loops can also have a 'copy'
 instruction, so detect those as well.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/MergeHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/MergeHelper.java
index c240daa740a7766b1541dc8aea73025cb9486ac7..d0158aba3e732e854494bfbe80b128f47a70fc24 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/MergeHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/MergeHelper.java
@@ -491,6 +491,18 @@ public class MergeHelper {
         stat.setIncExprent(holder.getInstance());
         preData.getExprents().remove(initExprents[0]);
         firstData.getExprents().remove(firstDoExprent);
+
+        if (initExprents[1] != null && initExprents[1].getLeft().type == Exprent.EXPRENT_VAR &&
+            holder.getInstance().type == Exprent.EXPRENT_VAR) {
+          VarExprent copy = (VarExprent)initExprents[1].getLeft();
+          VarExprent inc = (VarExprent)holder.getInstance();
+          if (copy.getIndex() == inc.getIndex() && copy.getVersion() == inc.getVersion() && !ExprentUtil.isVarReferenced(inc, stat, copy)) {
+            preData.getExprents().remove(initExprents[1]);
+            initExprents[1].getRight().addBytecodeOffsets(initExprents[1].bytecode);
+            initExprents[1].getRight().addBytecodeOffsets(stat.getIncExprent().bytecode);
+            stat.setIncExprent(initExprents[1].getRight());
+          }
+        }
         return true;
       }
       else if (initExprents[0] != null && initExprents[1] != null && firstDoExprent != null) {
