From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Mon, 19 Oct 2015 18:50:25 -0700
Subject: [PATCH] Fixed invalid reference to ASM.. Oops..


diff --git a/src/org/jetbrains/java/decompiler/main/rels/NestedClassProcessor.java b/src/org/jetbrains/java/decompiler/main/rels/NestedClassProcessor.java
index 1e69d8d4c01f89a1703a8e40424e8cb3c4ba62d3..3975d8cf93f8a2ce22b57ca3a666f3be77b5dcb4 100644
--- a/src/org/jetbrains/java/decompiler/main/rels/NestedClassProcessor.java
+++ b/src/org/jetbrains/java/decompiler/main/rels/NestedClassProcessor.java
@@ -43,8 +43,6 @@ import org.jetbrains.java.decompiler.struct.gen.MethodDescriptor;
 import org.jetbrains.java.decompiler.struct.gen.VarType;
 import org.jetbrains.java.decompiler.util.InterpreterUtil;
 
-import com.sun.xml.internal.ws.org.objectweb.asm.Opcodes;
-
 import java.util.*;
 import java.util.Map.Entry;
 
@@ -135,7 +133,7 @@ public class NestedClassProcessor {
       return;
     }
 
-    final int STATIC_FINAL_SYNTHETIC = Opcodes.ACC_STATIC | Opcodes.ACC_FINAL | Opcodes.ACC_SYNTHETIC;
+    final int STATIC_FINAL_SYNTHETIC = CodeConstants.ACC_STATIC | CodeConstants.ACC_FINAL | CodeConstants.ACC_SYNTHETIC;
     Set<String> potentialFields = new HashSet<String>();
     for (StructField fd : node.classStruct.getFields()) {
       if ((fd.getAccessFlags() & STATIC_FINAL_SYNTHETIC) == STATIC_FINAL_SYNTHETIC && "[I".equals(fd.getDescriptor())) {
