From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Mon, 19 Oct 2015 18:51:10 -0700
Subject: [PATCH] Fixed potential NPE in lvt propogation.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
index 0427e3b329d4599d42ef5f8ac35fd84581c5f8c6..aee11f60aab5e1f790a607e78675ba23f774fb47 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
@@ -434,7 +434,10 @@ public class VarDefinitionHelper {
     }
 
     for (VarType var : md.params) {
-      types.put(new VarVersionPair(index, 0), varproc.getLVT().getCandidates(index).get(0));
+      List<LVTVariable> vars = varproc.getLVT().getCandidates(index);
+      if (vars != null) {
+        types.put(new VarVersionPair(index, 0), vars.get(0));
+      }
       index += var.stackSize;
     }
 
