From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Mon, 19 Oct 2015 18:51:39 -0700
Subject: [PATCH] Fixed nested inner generic type printing.


diff --git a/src/org/jetbrains/java/decompiler/struct/gen/generics/GenericType.java b/src/org/jetbrains/java/decompiler/struct/gen/generics/GenericType.java
index 9ad5a1fe83d6f2e43f257bad5982e1b524e9d3af..9b111e572b474328818f044002d50dc22e946c5c 100644
--- a/src/org/jetbrains/java/decompiler/struct/gen/generics/GenericType.java
+++ b/src/org/jetbrains/java/decompiler/struct/gen/generics/GenericType.java
@@ -280,7 +280,7 @@ public class GenericType extends VarType {
       clsName = DecompilerContext.getImportCollector().getShortName(value.replace('/', '.'));
     }
     else if (parent.isGeneric()) {
-      clsName = ((GenericType)parent).getCastName() + "." + value.substring(value.lastIndexOf('.') + 1);
+      clsName = ((GenericType)parent).getCastName() + "." + value.substring(parent.value.length() + 1);
     }
     else {
       clsName = DecompilerContext.getImportCollector().getShortName(parent.value.replace('/', '.')) + "." + value.substring(value.lastIndexOf('.') + 1);
