From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cpw <cpw@weeksfamily.ca>
Date: Tue, 20 Oct 2015 19:20:44 -0400
Subject: [PATCH] Snapshots because we might be doing a bit more work yet


diff --git a/build.gradle b/build.gradle
index 5e14c2528a26cc50323d2d408cd16c2506d6b6c1..bc211f2219fbcef72a0acca7b79f5c41fcf0affa 100644
--- a/build.gradle
+++ b/build.gradle
@@ -3,7 +3,7 @@ apply plugin: 'idea'
 apply plugin: 'maven'
 apply plugin: 'java'
 
-version = "2.0"
+version = "2.0-SNAPSHOT"
 group = "net.minecraftforge" //not sure if we shoulkd publish under this name, but it's our fork so meh
 archivesBaseName = "fernflower"
 
