From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Thu, 22 Oct 2015 16:14:59 -0700
Subject: [PATCH] Fixed switch cases leaking variables incorrectly.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
index 520b951dd96e4cc70a1800af58aa7a9b43ac1d9f..726edd7aeab41647414952e0a0a6c8cb52674816 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
@@ -648,6 +648,9 @@ public class VarDefinitionHelper {
               if (obj == stat.getFirst()) {
                 leaked.putAll(leaked_n); //First is outside the scope so leak!
               }
+              else {
+                leaked_n.clear();
+              }
             }
             else if (stat.type == Statement.TYPE_TRYCATCH ||
                      stat.type == Statement.TYPE_CATCHALL) {
