From cf2bb43998a31ffa206193b9f0661df00bfbb6c0 Mon Sep 17 00:00:00 2001
From: Lex Manos <LexManos@gmail.com>
Date: Thu, 22 Oct 2015 16:14:59 -0700
Subject: [PATCH 103/122] Fixed switch cases leaking variables incorrectly.

---
 .../modules/decompiler/vars/VarDefinitionHelper.java           | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
index 520b951..726edd7 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/vars/VarDefinitionHelper.java
@@ -648,6 +648,9 @@ public class VarDefinitionHelper {
               if (obj == stat.getFirst()) {
                 leaked.putAll(leaked_n); //First is outside the scope so leak!
               }
+              else {
+                leaked_n.clear();
+              }
             }
             else if (stat.type == Statement.TYPE_TRYCATCH ||
                      stat.type == Statement.TYPE_CATCHALL) {
-- 
2.21.0.windows.1

