From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cpw <cpw@weeksfamily.ca>
Date: Sun, 25 Oct 2015 14:26:33 -0400
Subject: [PATCH] Fix NPE


diff --git a/test/org/jetbrains/java/decompiler/MinecraftDecompilationTest.java b/test/org/jetbrains/java/decompiler/MinecraftDecompilationTest.java
index 0d68e8886c6f6584c4514e9b95f320fb05b9e941..2ae6da55e0d8ecb2fbdb39a6e42d1f179ac32203 100644
--- a/test/org/jetbrains/java/decompiler/MinecraftDecompilationTest.java
+++ b/test/org/jetbrains/java/decompiler/MinecraftDecompilationTest.java
@@ -177,7 +177,7 @@ public class MinecraftDecompilationTest {
             set.add(md5digest);
             System.out.println("Orig md5:"+md5s.get(fname));
             System.out.println("Variant md5: "+md5digest);
-            writeFile(expected,fname,outRoot);
+            if (expected != null) writeFile(expected,fname,outRoot);
             writeFile(found,fname+"."+md5digest,outRoot);
         } else if (!md5digest.equals(md5s.get(fname))) {
           System.out.println("Existing Variant: " + fname);
