From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cpw <cpw@weeksfamily.ca>
Date: Sun, 8 Nov 2015 13:28:24 -0500
Subject: [PATCH] Add the index to the param renamer, needed for sane param
 name generation..


diff --git a/src/org/jetbrains/java/decompiler/main/ClassWriter.java b/src/org/jetbrains/java/decompiler/main/ClassWriter.java
index 1338a42a36f02f479a7efbb280f2826a2948bb03..25d01e38f0c006411e0a6c5c053b16d188b0c9a1 100644
--- a/src/org/jetbrains/java/decompiler/main/ClassWriter.java
+++ b/src/org/jetbrains/java/decompiler/main/ClassWriter.java
@@ -731,7 +731,7 @@ public class ClassWriter {
             buffer.append(' ');
             String parameterName = methodWrapper.varproc.getVarName(new VarVersionPair(index, 0));
             if ((flags & (CodeConstants.ACC_ABSTRACT | CodeConstants.ACC_NATIVE)) != 0) {
-                parameterName = methodWrapper.methodStruct.renamer.renameAbstractParameter(parameterName);
+                parameterName = methodWrapper.methodStruct.renamer.renameAbstractParameter(parameterName, index);
             }
             buffer.append(parameterName == null ? "param" + index : parameterName); // null iff decompiled with errors
 
diff --git a/src/org/jetbrains/java/decompiler/main/IdentityRenamerFactory.java b/src/org/jetbrains/java/decompiler/main/IdentityRenamerFactory.java
index bc29407ea2503b16a1eff3b49a6579189eccd14f..e8c33402d6b7e0c07dbe2aacf73c2f8f500fc8c3 100644
--- a/src/org/jetbrains/java/decompiler/main/IdentityRenamerFactory.java
+++ b/src/org/jetbrains/java/decompiler/main/IdentityRenamerFactory.java
@@ -14,7 +14,7 @@ public class IdentityRenamerFactory implements IVariableNamingFactory, IVariable
   }
 
   @Override
-  public String renameAbstractParameter(String abstractParam) {
+  public String renameAbstractParameter(String abstractParam, int index) {
     return abstractParam;
   }
 
diff --git a/src/org/jetbrains/java/decompiler/main/extern/IVariableNameProvider.java b/src/org/jetbrains/java/decompiler/main/extern/IVariableNameProvider.java
index 9d6b1a5cdff732d251e078750b2a26243f93463e..26f10715a8d594f00480a80e9c94c8b3bf011998 100644
--- a/src/org/jetbrains/java/decompiler/main/extern/IVariableNameProvider.java
+++ b/src/org/jetbrains/java/decompiler/main/extern/IVariableNameProvider.java
@@ -6,6 +6,6 @@ import org.jetbrains.java.decompiler.modules.decompiler.vars.VarVersionPair;
 
 public interface IVariableNameProvider {
   public Map<VarVersionPair,String> rename(Map<VarVersionPair,String> variables);
-  public String renameAbstractParameter(String abstractParam);
+  public String renameAbstractParameter(String abstractParam, int index);
   public void addParentContext(IVariableNameProvider renamer);
 }
diff --git a/src/org/jetbrains/java/decompiler/util/JADNameProvider.java b/src/org/jetbrains/java/decompiler/util/JADNameProvider.java
index f2d2e3baf14f3004b1d377b9cb7a882741885d23..b588cec7b138dfffeb1d57a11aeb7e4b41fa5a38 100644
--- a/src/org/jetbrains/java/decompiler/util/JADNameProvider.java
+++ b/src/org/jetbrains/java/decompiler/util/JADNameProvider.java
@@ -179,7 +179,7 @@ public class JADNameProvider implements IVariableNameProvider {
     }
 
     @Override
-    public String renameAbstractParameter(String abstractParam) {
+    public String renameAbstractParameter(String abstractParam, int index) {
       return abstractParam;
     }
 
