From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cpw <cpw@weeksfamily.ca>
Date: Sun, 8 Nov 2015 16:12:19 -0500
Subject: [PATCH] Early set the logger so that it can be used during init
 context without NPEs.


diff --git a/src/org/jetbrains/java/decompiler/main/DecompilerContext.java b/src/org/jetbrains/java/decompiler/main/DecompilerContext.java
index d66145aef50938051b6a21f5d089b40cf9e8ef9f..22ab9e0fd0f4571f360492221a619f9419d8812a 100644
--- a/src/org/jetbrains/java/decompiler/main/DecompilerContext.java
+++ b/src/org/jetbrains/java/decompiler/main/DecompilerContext.java
@@ -68,7 +68,8 @@ public class DecompilerContext {
       try {
         currentContext.get().renamerFactory = Class.forName((String) DecompilerContext.getProperty(RENAMER_FACTORY)).asSubclass(IVariableNamingFactory.class).newInstance();
       } catch (Exception e) {
-        getLogger().writeMessage("Error loading renamer factory class", e);
+        e.printStackTrace();
+        System.err.println("Error loading renamer factory class");
       }
     }
     if (DecompilerContext.getNamingFactory() == null) {
diff --git a/src/org/jetbrains/java/decompiler/main/Fernflower.java b/src/org/jetbrains/java/decompiler/main/Fernflower.java
index 91e70d3132095281338593d2166eedbcbd09e5d6..8319da8a08838b77508269bf44fe2209bcf4a328 100644
--- a/src/org/jetbrains/java/decompiler/main/Fernflower.java
+++ b/src/org/jetbrains/java/decompiler/main/Fernflower.java
@@ -41,6 +41,7 @@ public class Fernflower implements IDecompiledData {
 
   public Fernflower(IBytecodeProvider provider, IResultSaver saver, Map<String, Object> options, IFernflowerLogger logger) {
     structContext = new StructContext(saver, this, new LazyLoader(provider));
+    DecompilerContext.setLogger(logger);
     DecompilerContext.initContext(options);
     DecompilerContext.setCounterContainer(new CounterContainer());
     DecompilerContext.setLogger(logger);
