From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cpw <cpw@weeksfamily.ca>
Date: Mon, 23 Nov 2015 11:56:24 -0500
Subject: [PATCH] Try and release the renamer after we're done with the class.
 Should reduce memory footprint somewhat.


diff --git a/src/org/jetbrains/java/decompiler/struct/StructClass.java b/src/org/jetbrains/java/decompiler/struct/StructClass.java
index ab0f118debc0d7dee08f79a12c3558909e652482..784860620c1c07884c28ac26d431396960312f19 100644
--- a/src/org/jetbrains/java/decompiler/struct/StructClass.java
+++ b/src/org/jetbrains/java/decompiler/struct/StructClass.java
@@ -147,6 +147,9 @@ public class StructClass extends StructMember {
     if (loader != null) {
       pool = null;
     }
+    for (StructMethod m : methods) {
+        m.renamer = null;
+    }
   }
 
   public ConstantPool getPool() {
diff --git a/src/org/jetbrains/java/decompiler/struct/StructMethod.java b/src/org/jetbrains/java/decompiler/struct/StructMethod.java
index 7e7f7580f5301ea94b4bf97440a471caa3f62c50..68f5427e95ffd45a0fd925b8912c79d07a999d17 100644
--- a/src/org/jetbrains/java/decompiler/struct/StructMethod.java
+++ b/src/org/jetbrains/java/decompiler/struct/StructMethod.java
@@ -61,7 +61,7 @@ public class StructMethod extends StructMember {
   private InstructionSequence seq;
   private boolean expanded = false;
   private VBStyleCollection<StructGeneralAttribute, String> codeAttributes;
-  public final IVariableNameProvider renamer;
+  public IVariableNameProvider renamer;
 
   public StructMethod(DataInputFullStream in, StructClass clStruct) throws IOException {
     classStruct = clStruct;
