From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cpw <cpw@weeksfamily.ca>
Date: Mon, 23 Nov 2015 14:01:47 -0500
Subject: [PATCH] Try to release resources *after* we're done, not before we've
 started


diff --git a/src/org/jetbrains/java/decompiler/main/ClassesProcessor.java b/src/org/jetbrains/java/decompiler/main/ClassesProcessor.java
index f22baa00bfcb99aa8e05ee323fdecacdf8cf177a..5feaf475b6e1100d40a907d1b42bb602f1db7984 100644
--- a/src/org/jetbrains/java/decompiler/main/ClassesProcessor.java
+++ b/src/org/jetbrains/java/decompiler/main/ClassesProcessor.java
@@ -354,6 +354,9 @@ public class ClassesProcessor {
 
     node.wrapper = null;
     node.classStruct.releaseResources();
+    for (StructMethod m : node.classStruct.getMethods()) {
+        m.renamer = null;
+    }
 
     for (ClassNode nd : node.nested) {
       destroyWrappers(nd);
diff --git a/src/org/jetbrains/java/decompiler/struct/StructClass.java b/src/org/jetbrains/java/decompiler/struct/StructClass.java
index 784860620c1c07884c28ac26d431396960312f19..ab0f118debc0d7dee08f79a12c3558909e652482 100644
--- a/src/org/jetbrains/java/decompiler/struct/StructClass.java
+++ b/src/org/jetbrains/java/decompiler/struct/StructClass.java
@@ -147,9 +147,6 @@ public class StructClass extends StructMember {
     if (loader != null) {
       pool = null;
     }
-    for (StructMethod m : methods) {
-        m.renamer = null;
-    }
   }
 
   public ConstantPool getPool() {
