From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cpw <cpw@weeksfamily.ca>
Date: Mon, 23 Nov 2015 14:27:08 -0500
Subject: [PATCH] Don't allocate a renamer for every single method read. That
 should cut down the memory quite a lot.


diff --git a/src/org/jetbrains/java/decompiler/struct/StructMethod.java b/src/org/jetbrains/java/decompiler/struct/StructMethod.java
index 68f5427e95ffd45a0fd925b8912c79d07a999d17..77fd181a17eb566fa0f0330c0c21fb08f787922f 100644
--- a/src/org/jetbrains/java/decompiler/struct/StructMethod.java
+++ b/src/org/jetbrains/java/decompiler/struct/StructMethod.java
@@ -80,7 +80,6 @@ public class StructMethod extends StructMember {
       attributes.addAllWithKey(codeAttributes);
       codeAttributes = null;
     }
-    this.renamer = DecompilerContext.getNamingFactory().createFactory(this);
   }
 
   @Override
@@ -121,6 +120,7 @@ public class StructMethod extends StructMember {
     if (containsCode && !expanded) {
       byte[] code = classStruct.getLoader().loadBytecode(this, codeFullLength);
       seq = parseBytecode(new DataInputFullStream(code), codeLength, classStruct.getPool());
+      this.renamer = DecompilerContext.getNamingFactory().createFactory(this);
       expanded = true;
     }
   }
