From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Curle <curle@gemwire.uk>
Date: Tue, 21 Apr 2020 14:00:05 +0000
Subject: [PATCH] Changed archive name to reflect the update to ForgeFlower


diff --git a/build.gradle b/build.gradle
index cf3d8ec6b2fe4260b788b46643e8a397e5d63e18..7dec9ec6fb04c6cd2081f2e452d072f7451ddcad 100644
--- a/build.gradle
+++ b/build.gradle
@@ -3,9 +3,9 @@ apply plugin: 'idea'
 apply plugin: 'maven'
 apply plugin: 'java'
 
-version = "2.0-SNAPSHOT"
+version = "0.2.3"
 group = "net.minecraftforge" //not sure if we shoulkd publish under this name, but it's our fork so meh
-archivesBaseName = "fernflower"
+archivesBaseName = "forgeflower"
 
 sourceCompatibility = '1.6'
 targetCompatibility = '1.6'
